### ISSUE TITLE
Can't define a divide and conquer multiplier

### ISSUE THREAD

**Issue by robdockins** [GaloisInc/cryptol#325]
TL;DR version: I think we need some way to test type naturals that extends the typing context with the result of the test.  For example, I'd like to write `if `(n <= 8) then f else g`  In a context were I don't know much about `n`; but where f assumes `(n <= 8)` and where g assumes `(n >= 9)`.

Longer version: I'm trying to define multiplication using Karatsuba's algorithm, which is a divide-and-conquer algorithm that splits multiplicands roughly in half on each recursive call.  My attempt to define this program (see below) fails to typecheck because Cryptol can't figure out that I bottom out into a base case before `n` is too small for the types to work out.

I don't see how to make progress on this (or similar divide-and-conquer algorithms) without some way to extend typing contexts with the result of a run-time test.

```
module Karatsuba where

kmult : {n} (fin n, n>=1) => [n] -> [n] -> [2*n]
kmult x y =
 if `n <= (8:[max (width 8) (width n)]) then
   (zero#x)*(zero#y)
 else
   splitmult`{n} x y


splitmult : {n, low, hi}
            (fin n, n>=2, fin low, fin hi, hi == n/2, low == n - hi)
         => [n] -> [n] -> [2*n]
splitmult x y =
      (zero # ac   # (zero:[2*low])) +
      (zero # abcd # (zero:[low])) +
      (zero # bd)
  where
   (a,b) = splitAt`{hi} x
   (c,d) = splitAt`{hi} y

   ac   = kmult`{hi} a c
   bd   = kmult`{low} b d
   abcd = kmult`{low} ((zero # a) + b) ((zero # c) + d)
```

```
Loading module Cryptol
Loading module Karatsuba

[error] at ./karatsuba.cry:4:1--8:21:
  Failed to validate user-specified signature.
    In the definition of 'kmult', at ./karatsuba.cry:4:1--4:6:
      for any type n
        fin n
        n >= 1
      =>
      n >= 2
        arising from
        use of expression splitmult
        at ./karatsuba.cry:8:4--8:13
```

**Comment by robdockins** [comment]
Inspired by a conversaion with @TomMD, I tried the following workaround which forces the argument to be at least a minimum size.  This is guaranteed to be a no-op at runtime, but manages to teach the typechecker that the size is large enough.

Unfortunately, this exposes a different problem.  Typechecking this program causes Cryptol and Z3 both (!) to burn CPU cycles for longer than I was willing to wait.

```
module Karatsuba where

kmult : {n} (fin n, n>=1) => [n] -> [n] -> [2*n]
kmult x y =
 if `n <= (8:[max (width 8) (width n)]) then
   (zero#x)*(zero#y)
 else
   take (splitmult (make_atleast`{2} x) (make_atleast`{2} y))

make_atleast : {n, m, a} (fin n, fin m) => [m]a -> [max n m]a
make_atleast x = zero#x

splitmult : {n, low, hi}
            (fin n, n>=2, fin low, fin hi, hi == n/2, low == n - hi)
         => [n] -> [n] -> [2*n]
splitmult x y =
      (zero # ac    # (zero:[2*low])) +
      (zero # ad_bc # (zero:[low])) +
      (zero # bd)
  where
   (a,b) = splitAt`{hi} x
   (c,d) = splitAt`{hi} y

   ac    = kmult`{hi} a c
   bd    = kmult`{low} b d
   ad_bc = (kmult`{low} ((zero # a) + b) ((zero # c) + d)) - ac - bd
```

**Comment by robdockins** [comment]
I'm happy to report that current Cryptol can type check a version of Karatsuba's algorithm using Tom's trick to coerce bit vectors to the desired length.  For posterity, here is a version of the algorithm I'm reasonably confident is correct.

```
module Karatsuba where

kmult : {limb,n} (fin n, fin limb, limb >= 6, n>=1) => [n] -> [n] -> [2*n]
kmult x y =
 if `n <= (demote`{limb, max (width limb) (width n)}) then
   (zero#x)*(zero#y)
 else
   take (splitmult`{limb} (make_atleast`{limb} x) (make_atleast`{limb} y))

make_atleast : {n, m, a} (fin n, fin m, Zero a) => [m]a -> [max n m]a
make_atleast x = zero#x

splitmult : {limb,n} (fin n, fin limb, limb >= 6, n>=6)
         => [n] -> [n] -> [2*n]

splitmult x y = (ac # bd) + (zero # ad_bc # (zero:[low]))
  where
   type hi  = n/2
   type low = n - hi

   (a,b) = splitAt`{hi} x
   (c,d) = splitAt`{hi} y

   ac : [2*hi]
   ac = kmult`{limb,hi} a c

   bd : [2*low]
   bd = kmult`{limb,low} b d

   a_b = (zext a) + (zext b)
   c_d = (zext c) + (zext d)

   ad_bc : [2*(low+1)]
   ad_bc = (kmult`{limb,low+1} a_b c_d) - (zext ac) - (zext bd)


property splitmult_correct_tiny (x:[9]) (y:[9]) =
  zext x * zext y == splitmult`{limb=7} x y

property splitmult_correct_small (x:[11]) (y:[11]) =
  zext x * zext y == splitmult`{limb=7} x y

property splitmult_correct_medium(x:[17]) (y:[17]) =
  zext x * zext y == splitmult`{limb=7} x y

property splitmult_correct_large (x:[59]) (y:[59]) =
  zext x * zext y == splitmult`{limb=7} x y

property splitmult_correct_huge (x:[511]) (y:[511]) =
  zext x * zext y == splitmult`{limb=63} x y
```

Although the question of using runtime evidence to enrich the typing context is interesting, it's rather complicated, and the existence of this workaround is sufficient that I think we should close this bug.

**Comment by kiniry** [comment]
That's a great results, @robdockins.
